core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -395,32 +395,54 @@ class ReplicaFetcherThread(name: String,
-          val epochStream = rlm.storageManager().fetchIndex(rlsMetadata, IndexType.LEADER_EPOCH)
-          val epochs = readLeaderEpochCheckpoint(epochStream, log.dir)
-
-          // Truncate the existing local log before restoring the leader epoch cache and producer snapshots.
-          truncateFullyAndStartAt(partition, offset)
-
-          log.maybeIncrementLogStartOffset(leaderLogStartOffset, LeaderOffsetIncremented)
-          epochs.foreach(epochEntry => {
-            log.leaderEpochCache.map(cache => cache.assign(epochEntry.epoch, epochEntry.startOffset))
-          })
-          info(s"Updated the epoch cache from remote tier till offset: $offset with size: ${epochs.size} for $partition")
-
-          // restore producer snapshot
-          val snapshotFile = Log.producerSnapshotFile(log.dir, offset)
-          Files.copy(rlm.storageManager().fetchIndex(rlsMetadata, IndexType.PRODUCER_SNAPSHOT), snapshotFile.toPath)
-          log.producerStateManager.reloadSegments()
-          log.loadProducerState(offset)
-          info(s"Built the leader epoch cache and producer snapshots from remote tier for $partition. " +
-            s"Active producers: ${log.producerStateManager.activeProducers.size}, LeaderLogStartOffset: $leaderLogStartOffset")
+          var rlsMetadata: Optional[RemoteLogSegmentMetadata] = Optional.empty()
+          val epoch = log.leaderEpochCache.flatMap(cache => cache.epochForOffset(leaderLocalLogStartOffset))
+          if (epoch.isDefined) {
+            rlsMetadata = rlm.fetchRemoteLogSegmentMetadata(partition, epoch.get, leaderLocalLogStartOffset)
+          } else {
+            // If epoch is not available, then it might be possible that this broker might lost its entire local storage.
+            // We may also have to build the leader epoch cache. To find out the remote log segment metadata for the
+            // leaderLocalLogStartOffset-1, start from the current leader epoch and subtract one to the epoch till
+            // finding the metdata.
+            var previousLeaderEpoch = currentLeaderEpoch
+            while (!rlsMetadata.isPresent && previousLeaderEpoch >= 0) {
+              rlsMetadata = rlm.fetchRemoteLogSegmentMetadata(partition, previousLeaderEpoch, leaderLocalLogStartOffset-1)
+              previousLeaderEpoch -= 1
+            }
+          }
+          if (rlsMetadata.isPresent) {
+            val epochStream = rlm.storageManager().fetchIndex(rlsMetadata.get(), IndexType.LEADER_EPOCH)
+            val epochs = readLeaderEpochCheckpoint(epochStream, log.dir)
+
+            // Truncate the existing local log before restoring the leader epoch cache and producer snapshots.
+            truncateFullyAndStartAt(partition, leaderLocalLogStartOffset)
+
+            log.maybeIncrementLogStartOffset(leaderLogStartOffset, LeaderOffsetIncremented)
+            epochs.foreach(epochEntry => {
+              log.leaderEpochCache.map(cache => cache.assign(epochEntry.epoch, epochEntry.startOffset))
```

Comment:
Say the leader's current epoch is 10, and the leaderLocalLogStartOffset is 100. Assume these are remote log segments
seg 85-90 (LE-8 with epoch 8's startOffset 85)
seg 90-95 (LE-9 with epoch 9's startOffset 90)
seg 95-105 (LE-10 with epoch 10's startOffset 100)
seg 105-115

It seems the logic here would locate the seg 95-105 (using the leaderLocalStartOffset) and cache the epochs
based on that segment. Any epoch history before that segment, i.e. LE-8 and LE-9 won't be retrieved by this follower. Is that understanding correct?

Reply:
@gitlw Are you ok with this explanation?

Reply:
That answers my question. Thanks!